File: 30-Day LeetCoding Challenge/C++/maxPathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = INT_MIN;
    int dfs(TreeNode* root) {
         if(root == NULL) {
            return 0;
        }
        int x = dfs(root->left);
        int y = dfs(root->right);
        ans = max(ans, x + y + root->val);
        return max(0, max(x, y) + root->val);
    }
    int maxPathSum(TreeNode* root) {
       dfs(root);
       return ans; 
    }
};

File: 30-Day LeetCoding Challenge/Python/maxPathSum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import sys
class Solution(object):
    ans = -sys.maxint - 1
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.dfs(root)
        return self.ans
    def dfs(self, root):
        if root == None:
            return 0
        x = self.dfs(root.left)
        y = self.dfs(root.right)
        self.ans = max(self.ans, x + y + root.val)
        return max(0, max(x,y) + root.val)